src/lib/storage/zero_suppression/simd_bp128/simd_bp128_packing.cpp

```diff
@@ -0,0 +1,438 @@
+    constexpr auto _32_bit = 32u;
+
+    // Number of integers that fit completely into the 32-bit sub-blocks
+    constexpr auto i_max = (_32_bit - carry_over) / bit_size;
+
+    for (auto i = 0u; i < i_max; ++i) {
+      const auto offset = carry_over + i * bit_size;
+      out_reg = _mm_and_si128(_mm_srli_epi32(in_reg, offset), mask);
+      _mm_storeu_si128(out++, out_reg);
+    }
+
+    constexpr auto next_offset = carry_over + i_max * bit_size;
+    constexpr auto num_first_bits = _32_bit - next_offset;
+
+    // Check if integers have been split across the 128-bit block boundary
+    if (next_offset < _32_bit) {
+      out_reg = _mm_srli_epi32(in_reg, next_offset);
+      in_reg = _mm_load_si128(in++);
+
+      out_reg = _mm_or_si128(out_reg, _mm_and_si128(_mm_slli_epi32(in_reg, num_first_bits), mask));
+      _mm_storeu_si128(out++, out_reg);
+    } else {
+      constexpr auto last_recursion = 1u;
+
+      // Only load another 128-bit block if it’s not the last recursion
+      if (remaining_recursions > last_recursion) {
+        in_reg = _mm_load_si128(in++);
+      }
+    }
+
+    // Calculate the new carry over
+    constexpr auto new_carry_over = next_offset < _32_bit ? bit_size - num_first_bits : 0u;
+    Unpack128Bit<bit_size, new_carry_over, remaining_recursions - 1u>{}(in, out, in_reg, out_reg, mask);
+  }
+};
+
+template <uint8_t bit_size, uint8_t carry_over>
+struct Unpack128Bit<bit_size, carry_over, 0u> {
+  void operator()(const __m128i* in, __m128i* out, __m128i& in_reg, __m128i& out_reg, const __m128i& mask) const {}
+};
+
+void unpack_128_zeros(uint32_t* out) {
+  static constexpr auto num_zeros = 128u;
+  std::fill(out, out + num_zeros, 0u);
+}
+
+}  // namespace
+
+void SimdBp128Packing::write_meta_info(const uint8_t* _in, uint128_t* _out) {
+  const auto in = reinterpret_cast<const __m128i*>(_in);
+  auto out = reinterpret_cast<__m128i*>(_out);
+
+  const auto meta_block_info_rgtr = _mm_loadu_si128(in);
+  _mm_store_si128(out, meta_block_info_rgtr);
+}
+
+void SimdBp128Packing::read_meta_info(const uint128_t* _in, uint8_t* _out) {
+  const auto in = reinterpret_cast<const __m128i*>(_in);
+  auto out = reinterpret_cast<__m128i*>(_out);
+
+  auto meta_info_block_rgtr = _mm_load_si128(in);
+  _mm_storeu_si128(out, meta_info_block_rgtr);
+}
+
+void SimdBp128Packing::pack_block(const uint32_t* _in, uint128_t* _out, const uint8_t bit_size) {
+  auto in = reinterpret_cast<const __m128i*>(_in);
+  auto out = reinterpret_cast<__m128i*>(_out);
+
+  auto in_reg = _mm_setzero_si128();
+  auto out_reg = _mm_setzero_si128();
+  const auto mask = _mm_set1_epi32((1ul << bit_size) - 1);
+
+  switch (bit_size) {
+    case 0u:
+      // No compression needed, since all values equal to zero.
+      return;
+
+    case 1u:
+      Pack128Bit<1u>{}(in, out, in_reg, out_reg, mask);
+      return;
+
+    case 2u:
+      Pack128Bit<2u>{}(in, out, in_reg, out_reg, mask);
+      return;
+
+    case 3u:
```

Comment:
🤢 
Can we at least reduce this to one line, using `clang-format off` if necessary?

Reply:
I think it’s okay. It’s in a file that is probably really going to be looked at.

Reply:
ok